src/uoapi/course/course_info.py

```diff
@@ -1,166 +1,133 @@
-import sys
-from time import sleep, perf_counter as pf
-import urllib
-import json
-import itertools as it
-
 import requests
-from bs4 import BeautifulSoup
-#import pandas as pd
-import regex as re
-
-from uoapi.course import patterns as pt
-from uoapi.course import Prereq
+from bs4 import (
+    BeautifulSoup,
+    Tag,
+)
 
-#requests.packages.urllib3.util.ssl_.DEFAULT_CIPHERS += ':DES-CBC3-SHA'
-#timetable_url = 'https://web30.uottawa.ca/v3/SITS/timetable/Search.aspx'
+from uoapi.course import (
+    patterns as pt,
+    utils,
+    Prereq,
+    parse,
+)
+from uoapi.course.models import (
+    Subject,
+    Course,
+)
 
 # Course Info Parameters
-course_url = 'https://catalogue.uottawa.ca/en/courses/'
+course_url = "https://catalogue.uottawa.ca/en/courses/"
 
 #############################################################################
 # COURSE INFO SCRAPING
 #############################################################################
 
-def _extract_codes(string, return_all = True):
-    '''
-    Returns course codes found in string; 
-    if multiple codes are found and return_all is False, then returns an invalid code
-    Used in get_subjects.ipynb ''' 
-    codes = list({x.group(0) for x in re.finditer(pt.code_re, string)})
-    if return_all or len(codes) == 1:
-        return codes
-    return 'XXX 0000'
-
-def _extract_credits(string):
-    '''
-    Searches string for a number of credits/units
-    (Assuming the string is the title of a course)
-    Used in get_subjects.ipynb
-    '''
-    credits = list({int(x.group(0).split(' ')[0].strip('(')) for x in re.finditer(pt.credit_re, string)})
-    if len(credits) == 1:
-        return credits
-    return [0]
-
-
-def scrape_subjects(url=course_url):
-    '''
-    Scrapes the list of subjects with links to their respective course catalogues from the uOttawa website
-    () -> pandas DataFrame with columns: Subject, Code, Link
-    Used in get_subjects.ipynb
-    '''
+
+def scrape_subjects(url: str = course_url):
+    """
+    Scrapes the list of subjects with links to their respective course
+    catalogues from the uOttawa website
+
+    Args:
+        url: The url to scrape from
+
+    Returns:
+        A list of Subject objects converted to dictionaries
+    """
     page = requests.get(url).text
+    soup = BeautifulSoup(page, "html.parser")
+
+    match content := soup.find("div", attrs={"class": "az_sitemap"}):
+        # BeautifulSoup doesn't provide type annotations so mypy doesn't
+        # play nice with this in a case statement
+        case Tag():  # type: ignore
+            pass
+        case _:
+            raise ValueError("Could not find div with class az_sitemap")
```

Comment:
I don't really care which, but I'm curious if there's a reason this is preferable over
```
content = soup.find("div", attrs={"class": "az_sitemap"}):
if not isinstance(content, Tag):
    raise ValueError("Could not find div with class az_sitemap")
```
?

Reply:
I think it's just cuz I'm in general pretty hesitant to use `isinstance` and other reflective type functions

Reply:
Fair enough! I mean, I assume `match` uses those mechanisms under the hood anyways, but I was just curious.